### ISSUE TITLE
CBD side effect doesn't apply to conditionally introduced dependency

### ISSUE THREAD

**Issue by magodo** [hashicorp/terraform#26775]
### Terraform Version

```
v0.13.2
```

### Terraform Configuration Files

According to [this reply](https://github.com/hashicorp/terraform/pull/24662#issuecomment-652500328), the CBD annotation could be used to acheive to update a depending resource first and destroy the depended/orphan resource afterwards.

E.g. Given following config:

```terraform
resource "demo_resource_foo" "magodo" {
  name = "magodo"
  job = "xxx"
}

resource "demo_resource_foo" "kinoko" {
  name = "kinoko"
  job = demo_resource_foo.magodo.job
  lifecycle {
    create_before_destroy = true
  }
}
```

If we change it to:

```terraform
resource "demo_resource_foo" "kinoko" {
  name = "kinoko"
  #job = demo_resource_foo.magodo.job
  lifecycle {
    create_before_destroy = true
  }
}
```

The expected provision order is to update `demo_resource_foo.kinoko` first, then deleting `demo_resource_foo.magodo`.

However, this doesn't apply to the case that the dependency is introduced conditionally. Given:

```terraform
variable "foo" {
  type = bool
}

resource "demo_resource_foo" "magodo" {
  count = var.foo == true? 1:0
  name = "magodo"
  job = "xxx"
}

resource "demo_resource_foo" "kinoko" {
  name = "kinoko"
  job = var.foo == true? demo_resource_foo.magodo[0].job : "yyy"
  lifecycle {
    create_before_destroy = true
  }
}
```

If we provision with `-var=foo=true` first, then provision with it set to `false`. The order is opposite as before, which is unexpected.

**Comment by jbardin** [comment]
Hi @magodo 

I'm not able to reproduce this with either the latest 0.13 or 0.14. 
In both cases, the `kinoko` resource is updated before `magado` is destroyed. 

Do you have a more complete example we can work with demonstrating this behavior?

Thanks!

**Comment by magodo** [comment]
@jbardin Following is the execution flow on my local env:

```shell
💤 tf apply -auto-approve -var=foo=true
demo_resource_foo.magodo[0]: Creating...
demo_resource_foo.magodo[0]: Creation complete after 0s [id=magodo]
demo_resource_foo.kinoko: Creating...
demo_resource_foo.kinoko: Creation complete after 0s [id=kinoko]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

💤 tf apply -auto-approve -var=foo=false
demo_resource_foo.magodo[0]: Refreshing state... [id=magodo]
demo_resource_foo.kinoko: Refreshing state... [id=kinoko]
demo_resource_foo.magodo[0]: Destroying... [id=magodo]
demo_resource_foo.magodo[0]: Destruction complete after 0s
demo_resource_foo.kinoko: Modifying... [id=kinoko]
demo_resource_foo.kinoko: Modifications complete after 0s [id=kinoko]

Apply complete! Resources: 0 added, 1 changed, 1 destroyed.

💤 tf version
Terraform v0.13.5
+ provider registry.terraform.io/magodo/demo v99.99.99
```

**Comment by jbardin** [comment]
Thanks @magodo 

I can reproduce this with 0.13, and have confirmed it is fixed in the 0.14 branch.

Thanks!
